### ISSUE TITLE
Adjustments design for Nextgen realization dataset and job workflow automation 

### ISSUE THREAD

**Issue by robertbartel** [NOAA-OWP/DMOD#251]
### The TL;DR:

1. DMOD has logic to automate forcing dataset selection, so a user doesn't have to know/provide details of forcing datasets when requesting jobs
2. DMOD does require a realization config to start a Nextgen job
3. A valid Nextgen realization config requires details on the forcing data, which for DMOD implies details of the forcing dataset, thus undermining **1.**

### Details

DMOD intends to intelligently determine certain things for a user, to relieve certain burdens.  One example is that DMOD will automatically find a suitable forcing dataset for a requested Nextgen job, assuming one is available, rather than requiring the user to specify one.  In the future, DMOD may produce such a dataset on-demand, deriving it from other existing datasets (e.g., one containing the appropriate data but not in a directly supported format).

Separately, in order for the user to request a Nextgen job execution workflow, DMOD currently requires a Nextgen realization config.  More specifically, since DMOD is designed to make data accessible via datasets, the request must specify a  `NGEN_REALIZATION_CONFIG` format dataset, presumed to contain a valid, finalized realization config.  

These two things conflict:  a valid realization config will including certain details about the forcings, but the intent is for the user to not need to know and provide details on the forcing dataset.

The primary problem is that, unless the particular forcing dataset is known, the correct forcing provider type settings for the realization config (e.g., _CsvPerFeature_) cannot be known.  Further, at least initially, there are also problems with the forcing `path` settings, as the path to datasets within the Docker worker containers is tied to the dataset name.

### Solutions

The only feasible solutions I see are:

1. do not require a complete realization config when a job is requested
2. dynamically edit the saved realization config during the steps of the workflow
3. a solution made possible by altering the format of Nextgen realization configurations and functionality of Nextgen

I personally think option **1** is most appropriate, but some discussion is probably needed.

#### Option 1

If To represent config details that are provided by the user during the request, an intermediate dataset format would be needed.  Note that it should still be possible to make requests based on existing realization config datasets, as can be done now.  This just leaves the user responsible for their semantic correctness.   

Doing that would require something like the following:

- [ ] Add intermediate dataset format _JOB_CONFIG_ (or something similar) as a representation of a partial realization configuration 
- [ ] Update messaging and job workflow start logic to accept either _JOB_CONFIG_ (with sufficient data) or _NGEN_REALIZATION_CONFIG_ for Nextgen job request
- [ ] Update the GUI (and perhaps the CLI) to create such datasets from user form input immediately prior to submitting applicable job requests
- [ ] Update job exec workflow (in data-service) to have _JOB_CONFIG_  format be satisfactory for realization config requirements at the _AWAITING_DATA_CHECK_ job status step
- [ ] Update job exec workflow to derive a realization config dataset from _JOB_CONFIG_ dataset and suitable forcing dataset at the _AWAITING_DATA_ job status step

#### Option 2

I don't particularly like option **2**.  Other than subsetting (or perhaps supersetting), the automated logic for manipulating  datasets in DMOD was only intended to adjust the format of user-provided data.  We can explore what this would take though.

#### Option 3

This is really a question that requires some feedback from the primary Nextgen team.

**Comment by robertbartel** [comment]
@hellkite500, @aaraney, @christophertubbs, @mattw-nws, this a bit of a large one, but I think this needs some discussion.  We can handle it async, or we can set up a meeting if everyone prefers.

**Comment by aaraney** [comment]
Thanks for writing this up, @robertbartel! I agree that options 1 is probably the most viable and suitable. In all cases, it seems we will need to produce valid realization (and calibration) configuration files, module `init_config` files, and any other associated files. Likewise, as you pointed out previously, challenges with `path` fields likely mean we will need to generate or interpolate `path` fields on the fly. Potentially based on deployment environment requirements.

With that in mind, i'm inclined to think utilizing an intermediate format from which we generate required configuration files is the way to go. We could potentially even stick a small configuration generation script in model exec images to inject, for example, linked library `path`s. I am not sold on, at least to start with, _storing_ these intermediate datasets as I have a hunch we will run into fun migration and backwards compatibility problems as we sort out what format works best for our use cases. At the same time, I think it's crucial that we can reproduce experiments and that includes configuration files. That might mean we decide to add additional metadata as a post model execution procedure too for example, specify the image and tag used to conduct the experiment. Sorry for talking in circles, im just on the fence.

Having had a look at a realization configuration file, if we go with option 1, it appears we may be able to base our intermediate dataset on the fields specified in the catchment realization or formulation sections. Given that we already capture start time, end time, and catchment ids in our current model exec messages, we should be able determine forcing needs using only the model names (assuming we can acquire `ngen` models' input and output vars) and configuration specific `variable_names_map`pings. It should be pretty straightforward to run a mark and sweep like algorithm to determine required formulation forcing fields. This does not hold for non-BMI forced realizations (I hope im using the right terminology, please correct me when im wrong) like `noah-owp-modular` which means we will likely need to register models' input source capability (`BMI` vs `InitConfig` vs `BMI | InitConfig`) and handle them differently. Granted I don't full understand the problem with determining forcings, particularly the hurdles `CsvPerFeature` introduces.

**Comment by aaraney** [comment]
Use subset of fields and or fields names from `ngen.config` `NgenRealization` and `Realization` as intermediate format. Namely, `time`, `formulations`, and `forcing`. We will also need to know which `hydrofabric`(s) are required. A broader discussion is required to discuss potentially relating hydrofabric's and their contents to formulations within the realization config. This is related, but beyond the scope of this issue.

Related, will need link catchment aggregated forcings' to hydrofabric's, likely at the dataset level.
